### DISCUSSION TITLE
What exactly we want to achieve in delay function?

### DISCUSSION THREAD

**Discussion by NitinSaxenait** [japaric/discovery#361] [Q&A] [answered]
Hi, I am going through this discovery book and I have successfully completed it till chapter 8 Leds Again. Now in chapter 9 Clock and timers i was going through these for loop delays using function delay but I am not getting what actually we want to achieve and what this loop is doing in function delay. **Can someone please help me with this?** I know we are not using delay as we were using it in the previous chapter we have to work with it differently. 
This is what I want to understand.
`#[inline(never)]
fn delay(_tim6: &tim6::RegisterBlock, ms: u16) {
    **const K: u16 = 3; // this value needs to be tweaked
    for _ in 0..(K * ms)** {
        aux9::nop()
    }
}`

Okay so I know why we are using nop here I just want to understand the working of constant k & for loop

**Comment by sirhcel** [comment, marked as answer]
Hello @NitinSaxenait, sections 9 is about shedding some light on the internals of a delay function to give you an impression how it could be implemented. It sketches two common approaches:

- By "counting instructions": looping which does not get optimized away by the compiler
- By "counting timer ticks": setting up a timer and polling its status until it expires

 Without any special measures, the compiler will optimize away an empty loop. But it offers a "blessed no operation" which won't: The function [`aux9::nop`](https://github.com/rust-embedded/cortex-m/blob/45e7ab72661a8823e98e8fba821f2addcea95d3f/src/asm.rs#L34) resolves to the compiler intrinsic `__nop` which serves exactly this purpose.

The constant `K` is the number of loop iterations you have to perform for a millisecond. So `K * ms` is the number of loop iterations for achieving the desired delay in milliseconds.

**Reply by NitinSaxenait** [reply]
@sirhcel thanks for the explanation. Now I know what this means.
